src/interfaces/records/handlers/records-write.ts

```diff
@@ -103,41 +110,69 @@ export class RecordsWriteHandler implements MethodHandler {
   };
 
   /**
-   * Further validation if data stream is undefined.
-   * NOTE: if data stream is not be provided but `dataCid` is provided,
-   * then we need to make sure that the existing record state is referencing the same data as the incoming message.
-   * Without this check will lead to unauthorized access of data (https://github.com/TBD54566975/dwn-sdk-js/issues/359)
+   * Puts the given data in storage unless tenant already has that data for the given recordId
+   *
+   * @throws {DwnError} with `DwnErrorCode.RecordsWriteMissingDataStream`
+   *                    if `dataStream` is absent AND the `dataCid` does not match the current data for the given recordId
+   * @throws {DwnError} with `DwnErrorCode.RecordsWriteMissingData`
+   *                    if `dataStream` is absent AND dataStore does not contain the given `dataCid`
+   * @throws {DwnError} with `DwnErrorCode.RecordsWriteDataCidMismatch`
+   *                    if the data stream resulted in a data CID that mismatches with `dataCid` in the given message
+   * @throws {DwnError} with `DwnErrorCode.RecordsWriteDataSizeMismatch`
+   *                    if `dataSize` in `descriptor` given mismatches the actual data size
    */
-  protected validateUndefinedDataStream(
-    dataStream: _Readable.Readable | undefined,
-    newestExistingMessage: TimestampedMessage | undefined,
-    incomingMessage: RecordsWriteMessage): void {
-    if (dataStream === undefined && incomingMessage.descriptor.dataCid !== undefined) {
-      if (newestExistingMessage?.descriptor.dataCid !== incomingMessage.descriptor.dataCid) {
+  public async putData(
+    tenant: string,
+    message: RecordsWriteMessage,
+    dataStream?: _Readable.Readable,
+    newestExistingMessage?: TimestampedMessage
+  ): Promise<void> {
+    let result: { dataCid: string, dataSize: number };
+    const messageCid = await Message.getCid(message);
+
+    if (dataStream === undefined) {
+      // dataStream must be included if message contains a new dataCid
+      if (newestExistingMessage?.descriptor.dataCid !== message.descriptor.dataCid) {
         throw new DwnError(
           DwnErrorCode.RecordsWriteMissingDataStream,
           'Data stream is not provided.'
         );
       }
+
+      const associateResult = await this.dataStore.associate(tenant, messageCid, message.descriptor.dataCid);
+      if (associateResult === undefined) {
+        // Question for reviewers: I think if this throws, it is indicative of an underlying bug in the data store.
+        //   If previous RecordsWrites have gone correctly, this shouldn't happen, right?
```

Comment:
I agree this condition should not happen under normal operation. Though I think it can also happen beyond a bug inside data store, under a contrived circumstance now that sync is at play:

1. a sync wrote a pruned initial `RecordsWrite` without data stream
1. A normal (not via sync) `RecordWrite` for the same record immediately follows referencing the same `dataCid` but also without data stream

So we could either keep the code you have to handle this case or alternatively we make `associate()` throw instead of returning `undefined`. Either case we should probably add a test to cover the scenario. The code you've got now is arguably still better.

Reply:
Added a test. Keeping the existing `associate()` behavior where it returns undefined for now.